concurrency-primer.tex

```diff
@@ -883,6 +883,94 @@ \subsection{Conclusion about lock-free}
+Since the comparison result indicates \monobox{v} unchanged, \monobox{v + 10} is swapped in.
+Here sleeping is only used to ensure the occurrence of ABA problem.
+In real world scenario, instead of sleeping, thread A could paused by being context switched for other tasks, including being preempted by higher priority tasks.
+This example seems harmless, but things can get nasty when atomic \textsc{RMW} operations are used in more complex data structures.
+
+In a broader context, the ABA problem occurs when changes occur between loading and comparing, but the comparing mechanism is unable to identify that the state of the target object is not the latest, yielding a false positive result.
+
+Back to thread pool example in \secref{rmw}, it contains ABA problem as well.
+In \monobox{worker} function, we have the thread trying to claim the job.
+
+\begin{ccode}
+    job_t *job = atomic_load(&thrd_pool->head->prev);
+    ...
+    while (!atomic_compare_exchange_weak(&thrd_pool->head->prev, &job,
+                                           job->prev))
+        ;
+\end{ccode}
+
+Consider the following scenario:
+\begin{enumerate}
+    \item There is only one job left.
+    \item Thread A loads the pointer to the job by \monobox{atomic\_load()}.
+    \item Thread A is preempted.
+    \item Thread B claims the job and successfully updates \monobox{thrd\_pool->head->prev}.
+    \item Thread B sets thread pool state to idle.
+    \item Main thread finishes waiting and adds more jobs.
+    \item Memory allocator reuses the recently freed memory as new jobs addresses.
+    \item Fortunately, the first added job has the same address as the one thread A held.
+    \item Thread A is back in running state. The comparison result is equal so it updates \monobox{thrd\_pool->head->prev} with the old \monobox{job->prev}, which is already a dangling pointer.
+    \item Another thread loads the dangling pointer from \monobox{thrd\_pool->head->prev}.
+\end{enumerate}
+
+Notice that even though \monobox{job->prev} is not loaded explicitly before comparison, compiler could place loading instructions before comparison. 
+At the end, the dangling pointer could either point to garbage or trigger segmentation fault.
+It could be even worse if nested ABA problem occurs in thread B.
+Also, the possibility to allocate a job with same address could be higher when using memory pool, meaning that more chances to have ABA problem occurred.
+In fact, pre-allocated memory should be used to achieve lock-free since \monobox{malloc} could have mutex involved in multi-threaded environment.
+
+Failure to recognize changed target object through comparison can result in stale information.
+The general concept of solving this problem involves adding more information to make different state distinguishable, and then making a decision on whether to act on the old state or retry with the new state.
```

Comment:
This statement was not clear as for the impact of ABA problem. Can you improve it?